build-info-extractor/src/main/java/org/jfrog/build/extractor/clientConfiguration/util/GitUtils.java

```diff
@@ -0,0 +1,143 @@
+package org.jfrog.build.extractor.clientConfiguration.util;
+
+import org.apache.commons.lang.StringUtils;
+import org.jfrog.build.api.Vcs;
+import org.jfrog.build.api.util.Log;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class GitUtils {
+    private static String CREDENTIALS_IN_URL_REGEXP = "((http|https):\\/\\/\\w.*?:\\w.*?@)";
+
+    private static File getDotGit(File file) {
+        if (file == null) {
+            return null;
+        }
+        File dotGit = new File(file, ".git");
+        if (dotGit.exists()) {
+            return dotGit;
+        }
+        return getDotGit(file.getParentFile());
+    }
+
+    /**
+     * Extract Vcs details from .git configuration.
+     */
+    public static Vcs extractVcs(File workingDir, Log log) throws IOException {
+        File dotGit = getDotGit(workingDir);
+        Vcs vcs = new Vcs();
+        vcs.setRevision(extractVcsRevision(dotGit, log));
+        vcs.setUrl(extractVcsUrl(dotGit, log));
+        return vcs;
+    }
+
+    private static String extractVcsUrl(File dotGit, Log log) throws IOException {
+        File pathToConfig = new File(dotGit, "config");
+
+        String originalUrl = "";
+        // Fetch url from config file
+        try (BufferedReader br = new BufferedReader(new FileReader(pathToConfig))) {
+            String line;
+            boolean isNextLineUrl = false;
+            while ((line = br.readLine()) != null) {
+                if (isNextLineUrl) {
+                    line = line.trim();
+                    if (line.startsWith("url")) {
+                        String[] split = line.split("=");
+                        if (split.length < 2) {
+                            throw new IOException("Failed to parse .git config");
+                        }
+                        originalUrl = split[1].trim();
+                        break;
+                    }
+                }
+                if (line.equals("[remote \"origin\"]")) {
+                    isNextLineUrl = true;
+                }
+            }
+        }
+        if (!originalUrl.endsWith(".git")) {
+            originalUrl += ".git";
+        }
+
+        String maskedUrl = maskCredentialsInUrl(originalUrl);
+        log.debug("Fetched url from git config: " + maskedUrl);
+        return maskedUrl;
+    }
+
+    static String maskCredentialsInUrl(String originalUrl) throws IOException {
+        Pattern pattern = Pattern.compile(CREDENTIALS_IN_URL_REGEXP);
```

Comment:
Compiling a reg exp is a pretty expensive operation. I suggest we do it only once, and save the compiled pattern.